subprojects/base-services/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -0,0 +1,292 @@
+        this.maxWorkerCount = maxWorkerCount;
+        this.coordinationService = coordinationService;
+        this.projectLockRegistry = new ProjectLockRegistry(coordinationService, parallelEnabled);
+        this.workerLeaseLockRegistry = new WorkerLeaseLockRegistry(coordinationService);
+        LOGGER.info("Using {} worker leases.", maxWorkerCount);
+    }
+
+    @Override
+    public WorkerLease getCurrentWorkerLease() {
+        Collection<? extends ResourceLock> operations = workerLeaseLockRegistry.getResourceLocks();
+        if (operations.isEmpty()) {
+            throw new IllegalStateException("No worker lease associated with the current thread");
+        }
+        return (DefaultWorkerLease) operations.toArray()[operations.size() - 1];
+    }
+
+    @Override
+    public WorkerLeaseCompletion maybeStartWorkerLease() {
+        Collection<? extends ResourceLock> workerLeases = workerLeaseLockRegistry.getResourceLocks();
+        if (workerLeases.isEmpty()) {
+            return getWorkerLease().start();
+        }
+        return new NoOpCompletion();
+    }
+
+    private synchronized DefaultWorkerLease getWorkerLease(LeaseHolder parent) {
+        int workerId = counter++;
+        Thread ownerThread = Thread.currentThread();
+        return workerLeaseLockRegistry.getResourceLock(parent, workerId, ownerThread);
+    }
+
+    @Override
+    public DefaultWorkerLease getWorkerLease() {
+        Collection<? extends ResourceLock> operations = workerLeaseLockRegistry.getResourceLocks();
+        LeaseHolder parent = operations.isEmpty() ? root : (DefaultWorkerLease) operations.toArray()[operations.size() - 1];
+        return getWorkerLease(parent);
+    }
+
+    @Override
+    public void stop() {
+        coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
+            @Override
+            public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
+                if (workerLeaseLockRegistry.hasOpenLocks()) {
+                    throw new IllegalStateException("Some worker leases have not been marked as completed.");
+                }
+                if (projectLockRegistry.hasOpenLocks()) {
+                    throw new IllegalStateException("Some project locks have not been unlocked.");
+                }
+                return FINISHED;
+            }
+        });
+    }
+
+    @Override
+    public ResourceLock getProjectLock(String gradlePath, String projectPath) {
+        return projectLockRegistry.getResourceLock(gradlePath, projectPath);
+    }
+
+    @Override
+    public void withoutProjectLock(Runnable runnable) {
+        final Collection<? extends ResourceLock> projectLocks = projectLockRegistry.getResourceLocks();
+        coordinationService.withStateLock(unlock(projectLocks));
+        try {
+            runnable.run();
```

Comment:
I've put some similar convenience methods on `WorkerLeaseService`.  They fit a little bit better there because the `withoutLocks()` method needs access to the worker lease registry and it promotes the use of `WorkerLeaseService` as the primary entry point for lock functionality.